### PR TITLE
fix(cli/permissions): Fix CWD and exec path leaks

### LINKED ISSUE #5627: Deno.cwd's read permission is ineffective
The `PermissionDenied` error leaks the current working directory, thus the permission check can be circumvented like so:
```js
// cwd.js
function unauthenticatedCwd() {
  try {
    Deno.cwd();
  } catch (e) {
    return e.message.split('"')[1];
  }
}
console.log(unauthenticatedCwd());
```
```
: deno run cwd.js 
/home/syrup/downloads
: deno --version
deno 1.0.0
v8 8.4.300
typescript 3.9.2
```

### LINKED ISSUE #5742: Permission denied when using Deno.watchFs() relative path
The following code will fail when running with `deno run --allow-read=./src app.ts`

```js
const watcher = Deno.watchFs("./src");
for await (const event of watcher) {
  console.log(">>>> event", event);
  // { kind: "create", paths: [ "/foo.txt" ] }
}
```

Error:

```sh
error: Uncaught PermissionDenied: read access to "./src", run again with the --allow-read flag
    at unwrapResponse ($deno$/ops/dispatch_json.ts:43:11)
    at Object.sendSync ($deno$/ops/dispatch_json.ts:72:10)
    at new FsWatcher ($deno$/ops/fs_events.ts:15:16)
    at Object.watchFs ($deno$/ops/fs_events.ts:38:10)
    at file:///home/pomatti/projects/dumb-deno/app.ts:1:22
```

However if i change the first line of the script to use an absolute path it works fine:

```js
const watcher = Deno.watchFs("/home/pomatti/projects/dumb-deno/src");
```

I guess `Deno.watchFs()` must allow relative paths when `--allow-read` is informed on startup?

### LINKED ISSUE #5786: listenTls allow-read whitelist not working
```ts
// test.ts
console.log(`cert: ${(await Deno.open('./server.cert')).rid}`);

Deno.listenTls({
  port: 8888,
  hostname: '0.0.0.0',
  certFile: './server.cert',
  keyFile: './server.key'
});

````
```
C:\workspace\deno>deno --version
deno 1.0.2
v8 8.4.300
typescript 3.9.2

C:\workspace\deno>deno run --allow-net=0.0.0.0:8888 --allow-read=. test.ts
cert: 3
error: Uncaught PermissionDenied: read access to "./server.cert", run again with the --allow-read flag
    at unwrapResponse ($deno$/ops/dispatch_json.ts:43:11)
    at Object.sendSync ($deno$/ops/dispatch_json.ts:72:10)
    at Object.listenTls ($deno$/ops/tls.ts:67:10)
    at Object.listenTls ($deno$/tls.ts:51:22)
    at file:///C:/workspace/deno/test.ts:3:16

C:\workspace\deno>deno run --allow-net=0.0.0.0:8888 --allow-read test.ts
cert: 3

C:\workspace\deno>
```

### PR DESCRIPTION
**Pull request by nayeemrmn**
Fixes #5627.
Fixes #5742.
Fixes #5786.

- [x] Add `permissions.check_read_blind(path, display)` for `Deno.cwd()` and `Deno.execPath()`. It allows you to get errors like:
  - `PermissionDenied: read access to <CWD>, run again with the --allow-read flag`
  - `PermissionDenied: read access to <exec_path>, run again with the --allow-read flag`
- [x] For FS functions that accept relative paths, only display the resolved path in permission messages if the CWD is allowed. Otherwise use the input path. Example for permission requests:
  - ```ts
    await Deno.permissions.request({ name: "read", path: "../abc" });
    ```
  - `deno run --unstable temp.ts`:
    `⚠️  Deno requests read access to "../abc". Grant? [g/d (g = grant, d = deny)]`
  - `deno run --unstable --allow-read=. temp.ts`:
    `⚠️  Deno requests read access to "/mnt/c/Users/Nayeem/projects/abc". Grant? [g/d (g = grant, d = deny)]`
- [x] For `Deno.realPath()`, check if the CWD is allowed when a relative path is given.
- [x] Add tests.

### CODE REVIEW

File: cli/permissions.rs [hunk 1]
```diff
@@ -161,12 +181,24 @@ impl Permissions {
...
+      &format!("read access to \"{}\"", display_path.display()),
       "--allow-read",
     )
   }
 
+  pub fn check_read_blind(
+    &self,
+    path: &Path,
+    display: &str,
+  ) -> Result<(), OpError> {
```
Comment (ry) [inline cli/permissions.rs:195]: Please add a triple slash comment here so future readers can easily distinguish between check_read

File: cli/state.rs [hunk 2]
```diff
@@ -468,6 +468,15 @@ impl State {
     self.borrow().permissions.check_read(path)
   }
 
+  #[inline]
+  pub fn check_read_blind(
+    &self,
+    path: &Path,
+    display: &str,
+  ) -> Result<(), OpError> {
```
Comment (ry) [inline cli/state.rs:476]: triple slash comment would be helpful here too

File: cli/tests/056_make_temp_file_write_perm.ts [hunk 3]
```diff
@@ -1,5 +1,6 @@
-const path = await Deno.makeTempFile({ dir: "./subdir/" });
-if (path.startsWith(Deno.cwd())) {
+const sep = Deno.build.os == "windows" ? "\\" : "/";
+const path = await Deno.makeTempFile({ dir: `.${sep}subdir` });
+if (path.startsWith(`.${sep}subdir${sep}`)) {
```
Comment (ry) [inline cli/tests/056_make_temp_file_write_perm.ts:3]: The branch for windows is rather unfortunate. Is it just to make the `startsWith` work?

Comment (nayeemrmn) [reply cli/tests/056_make_temp_file_write_perm.ts]: Cleaned up.

**Comment by ry** [comment]
@nayeemrmn Gosh - sorry for the slow review again. This looks good to me. Could you please draft a commit message that I can use when squashing this?

**Comment by nayeemrmn** [comment]
Commit message:
```
fix(cli/permissions): Fix CWD and exec path leaks

- fix: Add permissions.check_read_blind() to check for read access for a
  directory while anonymizing it in error messages.
- fix: For ops which affect relative paths, only display the resolved
  path in permission messages if the CWD is allowed. Otherwise display
  the input path.
- fix: Deno.realPath(), check if the CWD is allowed when a relative path
  is given.
- refactor: Factor out all path resolution and normalization to
  cli/permissions.rs. This includes paths given as whitelists as well as
  op arguments.
```

**Pull request merged**